ddl/ddl_worker.go

```diff
@@ -578,6 +678,169 @@ func chooseLeaseTime(t, max time.Duration) time.Duration {
+							return errors.Trace(err)
+						}
+						var currentTable table.PhysicalTable
+						if tbl, ok := tbl.(table.PartitionedTable); ok {
+							currentTable = tbl.GetPartition(physicalID)
+						} else {
+							currentTable = tbl.(table.PhysicalTable)
+						}
+						startHandle, _, _, runner, status, subTaskProcessedCount, startTime, err := t.GetDDLSubTaskReorgInfo(job.ID, taskID, currentTable.Meta().IsCommonHandle)
+						costTime := model.TSConvert2Time(txn.StartTS()).Unix() - startTime
+						switch status {
+						case meta.Reorganized:
+							lastLinearTaskID = taskID
+							reorganizeTaskNum++
+							break
+						case meta.Success:
+							if (taskID-lastLinearTaskID) == int64(1) || (taskID == int64(0)) {
+								isLinear = true
+							}
+							lastLinearTaskID = taskID
+							totalAddedCount = totalAddedCount + subTaskProcessedCount
+							job.SetRowCount(totalAddedCount)
+							err = t.SetReorgSubTaskStatus(job.ID, taskID, meta.Reorganized)
+							if err != nil {
+								return errors.Trace(err)
+							}
+							logutil.Logger(w.logCtx).Info("[ddl] DDL subTask success", zap.String("job", job.String()),
+								zap.String("subTaskId", strconv.FormatInt(taskID, 10)),
+								zap.String("processedCount", strconv.FormatInt(subTaskProcessedCount, 10)),
+								zap.String("costTime", strconv.FormatInt(costTime, 10)),
+								zap.String("runner", runner))
+							break
+						case meta.Failed:
+							err = t.SetReorgSubTaskStatus(job.ID, taskID, meta.Unclaimed)
+							if err != nil {
+								return errors.Trace(err)
+							}
+							err = t.SetReorgSubTaskRunner(job.ID, taskID, meta.RunnerEmptyStr)
+							if err != nil {
+								return errors.Trace(err)
+							}
+							logutil.Logger(w.logCtx).Info("[ddl] DDL subTask failed", zap.String("job", job.String()),
+								zap.String("subTaskId", strconv.FormatInt(taskID, 10)),
+								zap.String("processedCount", strconv.FormatInt(subTaskProcessedCount, 10)),
+								zap.String("costTime", strconv.FormatInt(costTime, 10)),
+								zap.String("runner", runner))
+							break
+						case meta.Running:
+							logutil.Logger(w.logCtx).Info("[ddl] DDL subTask running", zap.String("job", job.String()),
+								zap.String("subTaskId", strconv.FormatInt(taskID, 10)),
+								zap.String("processedCount", strconv.FormatInt(subTaskProcessedCount, 10)),
+								zap.String("costTime", strconv.FormatInt(costTime, 10)),
+								zap.String("runner", runner))
+							//TODO try to set a proper time to wait for subTask running
+							waitTime := int64(30)
+							if costTime > waitTime {
+								err = t.SetReorgSubTaskStatus(job.ID, taskID, meta.Failed)
+							}
+							break
+						}
+						if err == nil {
+							if isLinear {
+								// update ReorgData for DDL job
+								ver, err := getValidCurrentVersion(d.store)
+								if err != nil {
+									return errors.Trace(err)
+								}
+								_, endHandle, err := getTableRange(d, currentTable, ver.Ver, job.Priority)
```

Comment:
I think we don't need to keep the reorg info for every subtask, it's too detailed.
for owner instance, we just care about whether the subtask is done and collect the addedcount.